### PR TITLE
[NCCL] Destructor Blocks on WorkNCCL Completion

### PR DESCRIPTION
**Pull request by osalpekar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44163 [NCCL] Add Environment Variable to guard Async Error Handling feature
* **#41054 [NCCL] Destructor Blocks on WorkNCCL Completion**
* #41053 [NCCL] WorkNCCL Helper Functions
* #41052 [NCCL] Abort Errored and Timed Out NCCL Communicators from Watchdog Thread
* #41051 [NCCL] Use cudaEventQuery to Poll for GPU operation errors
* #41050 [NCCL] Timeout Loop Thread for Async Error Handling

**This Commit:**
ProcessGroupNCCL destructor now blocks until all WorkNCCL objects have either been aborted or completed and removed from the work vector.

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D22054298](https://our.internmc.facebook.com/intern/diff/D22054298/)

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -623,6 +640,13 @@ void ProcessGroupNCCL::workCleanupLoop() {
         ++it;
       }
     }
+
+    if (workList_.empty()) {
+      // Notify the main thread if it is blocked in the shutdown sequence,
+      // waiting for the work vector to become empty.
+      lock.unlock();
+      workVectorCV_.notify_one();
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:648]: Do we really need to do this? Wouldn't this automatically abort when `terminateProcessGroup_` is set to True? Or are we referring to some other thread here?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -447,6 +447,23 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
...
+  // workCleanup Thread to be scheduled again.
+  for (auto it = workList_.begin(); it != workList_.end();
+       /* no increment*/) {
+    auto& work = *it;
+    if (work->isCompleted()) {
+      it = workList_.erase(it);
+    } else {
+      ++it;
+    }
+  }
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:462]: Why do we need to perform this explicit cleanup? Once the destructor completes, wouldn't `workList_` automatically be freed?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -628,6 +647,13 @@ void ProcessGroupNCCL::workCleanupLoop() {
         ++it;
       }
     }
+
+    if (workList_.empty()) {
+      // Notify the main thread if it is blocked in the shutdown sequence,
+      // waiting for the work vector to become empty.
+      lock.unlock();
+      workListCV_.notify_one();
+    }
```
Comment (jiayisuse) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:656]: Seems destructor is not waiting for workList to be empty(). Wondering if somewhere else is waiting for this? or we can just remove?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 4]
```diff
@@ -452,6 +452,25 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
...
+    auto& work = *it;
+    if (work->isCompleted()) {
+      it = workList_.erase(it);
+    } else {
+      ++it;
+    }
+  }
+  // Wait for workList_ to become empty before proceeding with shutdown.
+  workListCV_.wait(lock, [&]() -> bool { return workList_.empty(); });
+  lock.unlock();
```
Comment (jiayisuse) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:472]: With above workList cleanup, we won't need to wait?

Comment (osalpekar) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:472]: We're still blocking to ensure the workList is empty, so the workCleanupThread will continue looping and removing works when they are completed.

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (osalpekar) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: Right after this code block, we are blocking in the destructor until the workList_ is empty (no unfinished collectives left). Ideally this cleanup would just be done in the workcleanup thread itself, but there was one corner case causing an issue here - Hongyi and I are continuing to investigate, and I'll create an issue regarding this.

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (osalpekar) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: This notifies the CV in the destructor that is waiting for the workList_ to become empty.

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (pritamdamania87) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: Do we need to block in the destructor until `workList_` is empty? How does removing completed items from workList_ help in the shutdown here?

Comment (osalpekar) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: If there are leftover WorkNCCL objects in workList_, this means there are incomplete collectives. So we block on the workList_ becoming empty to ensure that all collectives have either been completed or errored out before we destruct ProcessGroupNCCL. Ideally, the workCleanupThread will just do all of the cleanup. However, when models contain a SyncBatchNorm layer, we find that this cleanup had to occur in the destructor. Hongyi (@jiayisuse ) and I have investigated this quite a bit, and I've created a follow-up issue (#44403). We should be able to deduplicate that explicit cleanup in the destructor and let the workCleanupThread handle it completely, and I'll continue to push this as a Better Engineering task.

**Pull request closed without merging**